### PR TITLE
feat: add sharpness metric

### PR DESCRIPTION
**Pull request by begumcig**
## Description

- Adds SharpnessMetric, which computes focus via the variance of the Laplacian (higher = sharper). 

- CPU‑side numpy/OpenCV implementation.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Additional Notes

You can use the `SharpnessMetric` directly or via the EvaluationAgent:
```
from pruna.evaluation.metrics.metric_sharpness import SharpnessMetric
from pruna.evaluation.task import Task
from pruna.evaluation.evaluation_agent import EvaluationAgent
from pruna.data.pruna_datamodule import PrunaDataModule

# 1) Direct use 
metric = SharpnessMetric(kernel_size=5)

inputs, gts = next(iter(my_dataloader))          # your B×C×H×W tensors
outputs = my_model(inputs)

metric.update(inputs, gts, outputs)
print("Sharpness:", metric.compute().result)     # higher = sharper

# 2) Via Task / EvaluationAgent
dm = PrunaDataModule.from_string("LAION256")
dm.limit_datasets(10)

task = Task(
    request=["sharpness"],        # or [SharpnessMetric(kernel_size=5)]
    datamodule=dm,
)
agent = EvaluationAgent(task=task)

results = agent.evaluate(my_model)
print(results[0])                 # MetricResult(name='sharpness', value=…)

### CODE REVIEW

File: src/pruna/evaluation/metrics/metric_sharpness.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+from pruna.evaluation.metrics.utils import SINGLE, get_call_type_for_single_metric, metric_data_processor
+from pruna.logging.logger import pruna_logger
+
+METRIC_SHARPNESS = "sharpness"
+
+
+@MetricRegistry.register(METRIC_SHARPNESS)
+class SharpnessMetric(StatefulMetric):
+    """
+    Laplacian‑variance image‑sharpness metric.
```
Comment (sharpenb) [inline src/pruna/evaluation/metrics/metric_sharpness.py:37]: Could we improve this description to be more informative fo rthe user? Maybe add the apper of the paper which introduces this metric. Explain what it measures

File: src/pruna/evaluation/metrics/metric_sharpness.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+    Parameters
+    ----------
+    *args : Any
+        Additional arguments.
+    kernel_size : int
+        The size of the kernel for the Laplacian operator. Default is 3.
```
Comment (sharpenb) [inline src/pruna/evaluation/metrics/metric_sharpness.py]: What is the impact of this parameter?

File: src/pruna/evaluation/metrics/metric_sharpness.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    call_type : str
+        The type of call to use for the metric. IQA metrics are only supported for single mode.
```
Comment (sharpenb) [inline src/pruna/evaluation/metrics/metric_sharpness.py]: It says IQA here.

File: src/pruna/evaluation/metrics/metric_sharpness.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+        if images.ndim != 4:
+            pruna_logger.error(f"Expected 4‑D tensor (B, C, H, W); got shape {tuple(images.shape)}")
+            raise
+
+        # Move to CPU OpenCV only works on numpy
+        imgs = images.detach().cpu()
+
+        # If range is 0‑1 → scale to 0‑255 for uint8
+        if imgs.max() <= 1.0:
```
Comment (sharpenb) [inline src/pruna/evaluation/metrics/metric_sharpness.py]: IT will not be able to process iamges which are full mix 0 or 1.

File: src/pruna/evaluation/metrics/metric_sharpness.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            imgs = imgs * 255.0
+        imgs = imgs.to(torch.uint8)
+
+        for img in imgs:
+            if img.shape[0] == 3:
+                img_gray = transforms_functional.rgb_to_grayscale(img.float()).squeeze(0).numpy().astype(np.uint8)
```
Comment (sharpenb) [inline src/pruna/evaluation/metrics/metric_sharpness.py]: This metric trasnforms check the sharpness on the gray image only?

File: src/pruna/evaluation/metrics/metric_sharpness.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    def compute(self) -> MetricResult:
+        """
+        Compute the mean sharpness score.
+
+        Returns
+        -------
+        MetricResult
+            The sharpness metric result.
+        """
+        if not self.scores:
```
Comment (sharpenb) [inline src/pruna/evaluation/metrics/metric_sharpness.py:158]: The logic is not clear here? Could we add some comment to explain why we have this if conditions? :)

File: tests/evaluation/test_sharpness.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    comparison_results = metric.compute().result
+
+    metric.reset()
+
+    # Calculate sharpness for ground truth
+    metric.update(x, gt, gt)
+    gt_results = metric.compute().result
+
+    # Both should be positive sharpness values (higher is better)
+    assert comparison_results > 0.0
```
Comment (sharpenb) [inline tests/evaluation/test_sharpness.py:47]: One test would be to check that for images that are more blurry the metric indeed decrease :)

**Review by sharpenb — COMMENTED** [review]
Very cool new metric! This is close to merge. Beyond my comments, a last thing would be to have/extend a tutorial with these introduced metric (it can be in another PR). This tutorial could take an image compute the metrics on it, and then perturb it it (e.g. by adding blur or other things mentioned in [this paper](https://github.com/hendrycks/robustness)). The goal would be to justify the relevance of these metrics :)

**Review by davidberenstein1957 — APPROVED** [review]

File: tests/evaluation/test_sharpness.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    assert gt_results > 0.0
+
+
+def test_sharpness_blurry_vs_sharp():
```
Comment (sdiazlor) [inline tests/evaluation/test_sharpness.py:51]: Can this test be parametrized to keep the standard as the rest?
